doc/reference/trd105-time.md

```diff
@@ -0,0 +1,455 @@
+require a compare be set at least N ticks in the future) but MUST NOT
+be less than `interval`.
+
+
+6 `Frequency` and `Ticks` Implementations
+=================================
+
+The time HIL provides four standard implementations of `Frequency`:
+
+```rust
+pub struct Freq16MHz;
+pub struct Freq1MHz;
+pub struct Freq32KHz;
+pub struct Freq16KHz;
+pub struct Freq1KHz;
+```
+
+The time HIL provides three standard implementaitons of `Ticks`:
+
+```rust
+pub struct Ticks24Bits(u32);
+pub struct Ticks32Bits(u32);
+pub struct Ticks64Bits(u64);
+```
+
+The 24 bits implementation is to support some Nordic Semiconductor
+nRF platforms (e.g. nRF52840) that only support a 24-bit counter.
+
+
+7 Capsules
+===============================
+
+The Tock kernel provides three standard capsules:
+
+  * `capsules::alarm::AlarmDriver` provides a system call driver for
+    an `Alarm`.
+  * `capsules::virtual_alarm` provides a set of
+    abstractions for virtualizing a single `Alarm` into many.
+  * `capsules::virtual_timer` provides a set of abstractions for
+    virtualizing a single `Alarm` into many `Timer` instances.
+
+8 Required Modules
+===============================
+
+A chip MUST provide an instance of `Alarm` with a `Frequency` of `Freq32KHz`
+and a `Ticks` of `Ticks32Bits`.
+
+A chip MUST provide an instance of `Time` with a `Frequency` of `Freq32KHz` and
+a `Ticks` of `Ticks64Bits`.
+
+A chip SHOULD provide an Alarm with a `Frequency` of `Freq1MHz` and a `Ticks`
+of `Ticks32Bits`.
+
+
+9 Implementation Considerations
+===============================
+
+This section describes implementation considerations for hardware
+implementations.
+
+The trickiest aspects of implementing the traits in this document relate
+to the `Alarm` trait and the semantics of how and when callbacks
+are triggered. In particular, if `set_alarm` indicates a time that has
+already passed, then the implementation should adjust it so that it
+will trigger very soon (rather than wait for a wrap-around).
+
+This is complicated by the fact that as the code is executing, the
+underlying counter continues to tick. Therefore an implementation must
+also be careful that this "very soon" time does not fall into the
+past. Furthermore, many instances of timer hardware requires that a
+compare value be some minimum number of ticks in the future. In
+practice, this means setting "very soon" to be a safe number of ticks
+in the future is a better implementation approach than trying to be
+extremely precise and inadvertently choosing too soon and then waiting
+for a wraparound.
+
+Pseudocode to handle these cases is as follows:
+
+```
+set_alarm(self, reference, dt):
+  now = now()
+  expires = reference.wrapping_add(dt)
+  if !now.within_range(reference, expired):
+    expires = now
+
+  if expires.wrapping_sub(now) < MIN_DELAY:
+    expires = now.wrapping_add(MIN_DELAY)
+
+  clear_alarm()
+  set_compare(expires)
+  enable_alarm()
+```
+
+10 Acknowledgements
```

Comment:
I'd suggest that we add a short section that describes the changes from 101.

Reply:
Added.